### PR TITLE
Historic ask to update

### PR DESCRIPTION
**Pull request by aylapear**
- update to deal with comments in shinyrems issues https://github.com/bcgov/shinyrems/issues/126 and https://github.com/bcgov/shinyrems/issues/118
- add option that the function asks if you want to download the newer version similar to 2yr/4yr option
- kept syntax of code similar to `get_ems_data()`

### CODE REVIEW

**Comment by ateucher** [comment]
Thanks! I should be able to have a look this week

File: R/ems_duckdb.R [hunk 1]
```diff
@@ -37,43 +42,86 @@ download_historic_data <- function(force = FALSE, ask = TRUE, dont_update = FALS
...
-  csv_file <- download_ems_data(url)
-  on.exit(unlink(csv_file), add = TRUE)
-
-  if (db_exists) {
-    file.remove(db_path)
-    write_db_path()
+  update <- FALSE
+  if (force || !db_exists) {
+    update <- TRUE # nocov
+  } else if (db_exists) {
```
Comment (ateucher) [inline R/ems_duckdb.R]: I think this logic has already been covered (just in reverse) on line 42. Can they be consolidated?

File: R/ems_duckdb.R [hunk 2]
```diff
@@ -37,43 +42,86 @@ download_historic_data <- function(force = FALSE, ask = TRUE, dont_update = FALS
...
+      ans <- readline(
+        paste0(
+          "Your version of the historic ems data is dated ",
+          cache_date,
+          " and there is a newer version available. Would you like to download it? (y/n)"
+        )
+      )
+      if (tolower(ans) == "y") update <- TRUE
+      # nocov end
+    }
```
Comment (ateucher) [inline R/ems_duckdb.R]: This block doesn't check for the value of `ask` so will run regardless of its value, and will give the user a prompt, which I don't think is the intention in the case of `ask = FALSE`?

File: R/ems_duckdb.R [hunk 3]
```diff
@@ -37,43 +42,86 @@ download_historic_data <- function(force = FALSE, ask = TRUE, dont_update = FALS
...
+        "You can access and subset it with the 'read_historic_data' function, or
         attach it as a remote data.frame with 'connect_historic_db()' and
-        'attach_historic_data()' which you can then query with dplyr")
-  invisible(db_path)
-  # nocov end
+        'attach_historic_data()' which you can then query with dplyr"
+      )
+      invisible(db_path)
+      # nocov end
+    }
```
Comment (ateucher) [inline R/ems_duckdb.R]: This code is fully duplicated from the block above, unnecessarily I think.

**Review by ateucher — CHANGES_REQUESTED** [review]
Hi, thanks for this. I think I'm a little unclear what the intent is with these changes...

For point # 1 (addressing https://github.com/bcgov/shinyrems/issues/126), does using the combination of `ask = FALSE, force = FALSE`, possibly in combination with `dont_update` already accomplish that?

I've left some specific comments, but I think if you could tell me what the desired functionality is that would be helpful!

Thanks, and let me know if there's anything I can do to assist.

**Comment by aylapear** [comment]
Sorry Andy got tied up in a couple things the last couple weeks but will review your comments and get back to you shortly. 

The intent of the changes is to make a better user experience in shiny rems. It was expressed that the historic data is now updated daily when it used to be monthly. When a user is using the app it prompts them to download the data each day now and so an option to not re-download the new data was desired similar to how the 2yr or 4yr datasets functions. If you had just downloaded it the day before you may not want to take the time to re-download the entire historic dataset. It seemed like the current options for the historic data set didn't meet those needs exactly as the other data sets do.

File: R/ems_duckdb.R (hunk 1 above)
Comment (aylapear) [reply R/ems_duckdb.R]: I took a step back and found a way to refactor this section so it reads more clearly and is consolidated.

File: R/ems_duckdb.R (hunk 2 above)
Comment (aylapear) [reply R/ems_duckdb.R]: I agree with your comments here and I miss interpreted the `ask` parameter as the definition says "should the function ask for your permission to cache data on your computer?" which I took to being only about the question where it asks to save a copy of the data on your local machine. I have updated the function so the question in relation to the version is wrapped up in the ask parameter.

File: R/ems_duckdb.R (hunk 3 above)
Comment (aylapear) [reply R/ems_duckdb.R]: Yes this was duplicated, I have refactored and simplified it so doesn't duplicate the code anymore

**Comment by aylapear** [comment]
Hey Andy,

Thank you for the comments as it made me take a step back and look at a better way of structuring the code. The code is organized a bit differently now so let me know if you have any further comments or suggestions. I was trying to be quick and match more of the structure of the 2yr/4yr code but it made things more confusing and bloated then just trying to structure it for the historic setup. 

This change greatly improves functionality for the shinyrems app as it allows the user to know how current their data is and whether it is worth updating it vs before you only had the option to update it if it was out of date. 
```
> shinyrems::run_ems_app("historic")
checking for historic data...
Your version of the historic ems data is dated 2022-02-16 05:57:00 and there is a newer version available. Would you like to download it? (y/n)n
Loading required package: shiny

Listening on http://127.0.0.1:5762
```

**Comment by ateucher** [comment]
Thanks very much @aylapear for the rethinking and refactoring! It is nice and clean now - I'm just waiting for the CI to finish and then will be happy to merge!

File: R/ems_duckdb.R [hunk 4]
```diff
@@ -1,79 +1,99 @@
...
       return(invisible(db_path))
     }
 
     if (dont_update) {
       warning("There is a newer version of the historic ems data ",
-        ", however you have asked not to update it by setting 'dont_update' to TRUE.")
+              ", however you have asked not to update it by setting 'dont_update' to TRUE.")
       return(invisible(db_path))
     }
+
```
Comment (ateucher) [inline R/ems_duckdb.R]:
```suggestion
# nocov start
```

File: R/ems_duckdb.R [hunk 5]
```diff
@@ -1,79 +1,99 @@
...
+    if (ask) {
+      ans <- readline(
+        paste0(
+          "Your version of the historic ems data is dated ",
+          cache_date,
+          " and there is a newer version available. Would you like to download it? (y/n)"
+        )
+      )
+      if (tolower(ans) == "n") return(invisible(db_path))
+    }
```
Comment (ateucher) [inline R/ems_duckdb.R:64]:
```suggestion
    }
    # nocov end
```

File: R/ems_duckdb.R [hunk 6]
```diff
@@ -1,79 +1,99 @@
...
-    stop_for_permission(paste0("rems would like to store a copy of the historic ems data at ",
-      db_path, ". Is that okay?"))
+    stop_for_permission(
+      paste0(
+        "rems would like to store a copy of the historic ems data at ",
+        db_path,
+        ". Is that okay?"
+      )
+    )
   }
```
Comment (ateucher) [inline R/ems_duckdb.R:77]:
```suggestion
  }
  # nocov end
```

**Review by ateucher — CHANGES_REQUESTED** [review]
Looks great @aylapear - just a couple of places to insert some codecov comments (since we can't test the interactive bits)

**Comment by aylapear** [comment]
@ateucher I added those no cov lines which I was hoping will fix the issue with the failing action due to the code coverage

**Review by ateucher — APPROVED** [review]

**Comment by ateucher** [comment]
All green! Thanks so much @aylapear for the nice work and your patience with my slow reviews 😄

**Pull request merged**